### PR TITLE
Variable Fixing Action

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Wraps the RadiallyFallingFloor struct in an Action.
Tests this Action using the new Action testing framework

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/VariableFixing/Actions.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/// - DataBox:
+///   - Metavariables::fixing_scheme::return_tags
+///   - Metavariables::fixing_scheme::argument_tags
+/// - ConstGlobalCache:
+///   - Metavariables::fixing_scheme::const_global_cache_tags
+///
+/// DataBox changes:
+/// - Adds: nothing
+/// - Removes: nothing
+/// - Modifies: DensityTag, PressureTag
```
Comment (wthrowe) [inline src/Evolution/VariableFixing/Actions.hpp]: What are these?

File: src/Evolution/VariableFixing/Actions.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+template <typename VariableFixer>
+struct ApplyVariableFixer {
+  // using fixing_scheme = typename Metavariables::fixing_scheme;
```
Comment (wthrowe) [inline src/Evolution/VariableFixing/Actions.hpp]: Leftover code?

File: src/Evolution/VariableFixing/Actions.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}  // namespace Parallel
+/// \endcond
+
+namespace Actions {
+/// \ingroup ActionsGroup
+/// \brief Fix the pressure and rest mass density if they are too low.
+///
+/// Uses:
+/// - DataBox:
+///   - Metavariables::fixing_scheme::return_tags
```
Comment (wthrowe) [inline src/Evolution/VariableFixing/Actions.hpp]: I think all these `Metavariables::fixing_scheme` references are suppose to be `VariableFixer`.

File: tests/Unit/Evolution/VariableFixing/Test_Actions.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <cstddef>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include "tests/Unit/TestingFramework.hpp"
```
Comment (wthrowe) [inline tests/Unit/Evolution/VariableFixing/Test_Actions.cpp]: Should be first.

File: tests/Unit/Evolution/VariableFixing/Test_Actions.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  using LocalAlgorithms =
+      typename ActionTesting::ActionRunner<Metavariables>::LocalAlgorithms;
+  using component = mock_component<Metavariables>;
+  using ActionRunner = ActionTesting::ActionRunner<Metavariables>;
+  using LocalAlgsTag =
+      typename ActionRunner::template LocalAlgorithmsTag<component>;
+  LocalAlgorithms local_algs{};
+  const DataVector x{-2.0, -1.0, 0.0, 1.0, 2.0};
+  const DataVector y{-2.0, -1.0, 0.0, 1.0, 2.0};
+  const DataVector z{-2.0, -1.0, 0.0, 1.0, 2.0};
```
Comment (wthrowe) [inline tests/Unit/Evolution/VariableFixing/Test_Actions.cpp:59]: Are these intentionally all the same?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/VariableFixing/Test_Actions.cpp (hunk 5 above)
Comment (mar-celine) [reply tests/Unit/Evolution/VariableFixing/Test_Actions.cpp:59]: Yes, but they can be changed if that's wanted. (This is also done to match the test done for the struct RadiallyFallingFloor.) If they should match is also something I'm open to discussion about.

File: tests/Unit/Evolution/VariableFixing/Test_Actions.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/Tags.hpp"
+#include "Evolution/Systems/GrMhd/Tags.hpp"
+#include "Evolution/VariableFixing/Actions.hpp"
+#include "Evolution/VariableFixing/RadiallyFallingFloor.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/ActionTesting.hpp"
+
+// IWYU pragma; no_forward_declare ApplyVariableFixer
```
Comment (wthrowe) [inline tests/Unit/Evolution/VariableFixing/Test_Actions.cpp]: `;` -> `:` and add namespace, although this line may become unnecessary once the other IWYU complaints are resolved.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Review by nilsdeppe — COMMENTED** [review]
I'm happy for you to squash this once @wthrowe is happy with it :)

File: tests/Unit/Evolution/VariableFixing/Test_Actions.cpp (hunk 6 above)
Comment (mar-celine) [reply tests/Unit/Evolution/VariableFixing/Test_Actions.cpp]: Sorry, I don't know what you mean by "add namespace" - are you referring to the anonymous namespace on line 25? Should it be named?

Comment (kidder) [reply tests/Unit/Evolution/VariableFixing/Test_Actions.cpp]: ApplyVariableFixer is in the namespace Actions

**Comment by kidder** [comment]
rebase when squashing...

**Review by wthrowe — DISMISSED** [review]
Squash, rebase.

**Review by nilsdeppe — DISMISSED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**